assets/src/edit-story/components/canvas/carrousel/index.js

```diff
@@ -30,12 +31,45 @@ const Page = styled.a`
 `;
 
 function Canvas() {
-	const { state: { pages, currentPageIndex }, actions: { setCurrentPageByIndex } } = useStory();
+	const { state: { pages, currentPageIndex }, actions: { setCurrentPageByIndex, arrangePage } } = useStory();
+	const getArrangeIndex = ( sourceIndex, dstIndex, position ) => {
+		// If the dropped element is before the dropzone index then we have to deduct
+		// that from the index to make up for the "lost" element in the row.
+		const indexAdjustment = sourceIndex < dstIndex ? -1 : 0;
+		if ( 'left' === position.x ) {
+			return dstIndex + indexAdjustment;
+		}
+		return dstIndex + 1 + indexAdjustment;
+	};
 	return (
 		<List>
-			{ pages.map( ( page, index ) => (
-				<Page key={ index } onClick={ () => setCurrentPageByIndex( index ) } isActive={ index === currentPageIndex } />
-			) ) }
+			{ pages.map( ( page, index ) => {
+				const onDrop = ( evt, position ) => {
+					const droppedEl = JSON.parse( evt.dataTransfer.getData( 'text' ) );
+					if ( ! droppedEl || 'page' !== droppedEl.type ) {
+						return;
+					}
+					const arrangedIndex = getArrangeIndex( droppedEl.index, index, position );
+					// Do nothing if the index didn't change.
+					if ( droppedEl.index !== arrangedIndex ) {
+						arrangePage( droppedEl.index, arrangedIndex );
+						setCurrentPageByIndex( arrangedIndex );
+					}
+				};
+				// @todo Create a Draggable component for setting data and setting "draggable".
+				const onDragStart = ( evt ) => {
+					const pageData = {
+						type: 'page',
+						index,
+					};
+					evt.dataTransfer.setData( 'text', JSON.stringify( pageData ) );
+				};
+				return (
+					<DropZone key={ index } onDrop={ onDrop } >
```

Comment:
So, a `DropZone` always wraps a page thumbnail? Is there no way to drop between the two thumbnail elements? I mean is it physically possible, if the thumbnails have margins?

Reply:
Yes, it's possible, currently, too. The thumbnails have margin currently as well, meaning there is no space between the two dropzones and dropping between the thumbnails will still result in dropping into a dropzone. In theory, it would be possible to add a separate dropzone element between the thumbnails instead but it seems to be easier to just wrap the related element and adjust it via styling as necessary.